modules/consentManagementUsp.js

```diff
@@ -66,42 +66,121 @@ function lookupUspConsent(uspSuccess, uspError, hookConfig) {
   }
 
   let callbackHandler = handleUspApiResponseCallbacks();
-  let uspapiCallbacks = {};
+  let uspApiCallbacks = {};
 
   // to collect the consent information from the user, we perform a call to USPAPI
   // to collect the user's consent choices represented as a string (via getUSPData)
-
   // the following code also determines where the USPAPI is located and uses the proper workflow to communicate with it:
-  // - use the USPAPI locator code to see if USP's located in the current window or an ancestor window. This works in friendly or cross domain iframes
-  // - if USPAPI is not found, the iframe function will call the uspError exit callback to abort the rest of the USPAPI workflow
-  // - try to call the __uspapi() function directly, otherwise use the postMessage() api
-  // find the CMP frame/window
-
-  try {
-    // try to call __uspapi directly
-    window.__uspapi('getUSPData', USPAPI_VERSION, callbackHandler.consentDataCallback);
-  } catch (e) {
-    // must not have been accessible, try using postMessage() api
-    let f = window;
-    let uspapiFrame;
-    while (!uspapiFrame) {
-      try {
-        if (f.frames['__uspapiLocator']) uspapiFrame = f;
-      } catch (e) { }
-      if (f === window.top) break;
-      f = f.parent;
+
+  // starting with the current window, traverse up through the ancestor windows until we get to window.top
+  // 1. __uspapi() exists on window
+  // 2. uspapi() exists on safeframe
+  // 3. look for locator __uspapiLocator in x-domain iframe
+  // 4. if current window is not the top window, and window.top is not friendly call window.top.postMessage()
+  let f = window;
+  let uspApi;
+  while (!uspApi) {
+    if (uspApiDefinedOnWindow(f, callbackHandler.consentDataCallback) ||
+      uspApiDefinedOnSafeframe(f, callbackHandler.consentDataCallback) ||
+      uspApiDefinedOnIframe(f, callbackHandler.consentDataCallback)) {
+      uspApi = f;
+    }
+    if (f === f.top) break;
+    f = f.parent;
+  }
+
+  if (f !== f.top && !inASafeFrame(f.top)) {
+    callUspApiWhileInIframe(f.top, 'getUSPData', callbackHandler.consentDataCallback);
+    uspApi = f;
+  }
```

Comment:
what is going on here?  if it was found in the while() loop above, uspapi() would already have been called, right?  Is this supposed to be one last effort, hoping it's in *window.top* if we didn't find it anywhere already?  If so, I don't think this does that.   Maybe it should be `if (!uspApi && f !== f.top) {` ?